File: src/Systems/Go/composite.go
package main

import "fmt"

type Component interface {
	Size() int
}

type FileLeaf struct {
	bytes int
}

func (f FileLeaf) Size() int { return f.bytes }

type FolderComposite struct {
	children []Component
}

func (f FolderComposite) Size() int {
	total := 0
	for _, child := range f.children {
		total += child.Size()
	}
	return total
}

func main() {
	var readme Component = FileLeaf{bytes: 2}
	var docs Component = FolderComposite{children: []Component{FileLeaf{bytes: 3}, FileLeaf{bytes: 5}}}
	var root Component = FolderComposite{children: []Component{readme, docs}}

	fmt.Printf("leaf=%d\n", readme.Size())
	fmt.Printf("docs=%d\n", docs.Size())
	fmt.Printf("root=%d\n", root.Size())
}

File: src/Systems/Swift/composite.swift
protocol Component {
    func size() -> Int
}

struct FileLeaf: Component {
    let bytes: Int
    func size() -> Int { bytes }
}

struct FolderComposite: Component {
    let children: [any Component]
    func size() -> Int { children.reduce(0) { $0 + $1.size() } }
}

let readme: any Component = FileLeaf(bytes: 2)
let docs: any Component = FolderComposite(children: [FileLeaf(bytes: 3), FileLeaf(bytes: 5)])
let root: any Component = FolderComposite(children: [readme, docs])

print("leaf=\(readme.size())")
print("docs=\(docs.size())")
print("root=\(root.size())")
